File: compiler/bench/combine.c
// expected: 184756
#include <stdio.h>
long comb_count = 0;

void combine(int n, int k, int start) {
    if (k == 0) { comb_count++; return; }
    if (n - start + 1 < k) return;
    for (int i = start + 1; i <= n; i++) {
        combine(n, k - 1, i);
    }
}

int main() {
    combine(20, 10, 0);
    printf("%ld\n", comb_count);
    return 0;
}

File: compiler/bench/combine.fs
\ expected: 184756
\ Generate all combinations C(20,10)

variable comb-count

: combine ( n k start -- ) recursive
  over 0= if 2drop drop 1 comb-count +! exit then
  over 2 pick - 2 pick + 1+ over > if 2drop drop exit then
  2 pick 1+ begin
    dup 3 pick <= while
    2 pick 3 pick 1- rot recurse
    1+
  repeat drop 2drop ;

: main
  0 comb-count !
  20 10 0 combine
  comb-count @ . cr ;
